### ISSUE TITLE
[macOS] Clicks don't always register on Big Sur

### ISSUE THREAD

**Issue by kartikeybhardwaj** [flutter/flutter#64961]
## Steps to Reproduce

1. Switch to dev channel and upgrade
2. Enable config for macos - `flutter config --enable-macos-desktop`
3. Create app - `flutter create myapp`
4. Update default template's class `_MyHomePageState` with the one below-
```
class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
            Expanded(
              child: InkWell(
                onTap: _incrementCounter,
                child: Text('Click me'),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

**Expected results:**
01. On click of `FloatingActionButton` counter should increments.
02. On tap of `InkWell` counter should increment.

**Actual results:**
01. On click of `FloatingActionButton` sometimes work and sometimes doesn't.
02. On tap of `InkWell` inside `Expanded` sometimes work and sometimes doesn't.

<details>
  <summary>Logs</summary>

Unable to attach output of `flutter run -d macos --verbose` as it is very huge.

Output of running `flutter doctor -v`

```
[✓] Flutter (Channel dev, 1.22.0-1.0.pre, on macOS 11.0 20A5354i, locale en-GB)
    • Flutter version 1.22.0-1.0.pre at /Users/kartoon/flutter
    • Framework revision ce40de69b7 (11 days ago), 2020-08-20 07:31:50 -0700
    • Engine revision 81027ab0cc
    • Dart version 2.10.0 (build 2.10.0-45.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/kartoon/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode-beta.app/Contents/Developer
    • Xcode 12.0, Build version 12A8189h
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (2 available)
    • ONEPLUS A6000 (mobile) • 3382e506 • android-arm64 • Android 10 (API 29)
    • macOS (desktop)        • macos    • darwin-x64    • macOS 11.0 20A5354i
    ! Error: kartoon's xr is not connected. Xcode will continue when kartoon's xr is connected. (code -13)
```

</details>

**Comment by darshankawar** [comment]
Hi @kartikeybhardwaj,
I tried the same and was able to see the counter incrementing everytime I tap on `Click me` text.

[64961.mov.zip](https://github.com/flutter/flutter/files/5154470/64961.mov.zip)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter: is fully installed. (Channel dev, 1.22.0-1.0.pre, on Mac OS X
    10.15.2 19C57, locale en-US)
[!] Android toolchain - develop for Android devices: is partially installed;
    more components are available. (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS: is fully installed. (Xcode 11.6)
[✓] Chrome - develop for the web: is fully installed.
[✓] Android Studio: is fully installed. (version 3.6)
[✓] Connected device: is fully installed. (4 available)
```

</details>

Can you provide a video / gif when it sometimes doesn't work ?
Thanks.

**Comment by kartikeybhardwaj** [comment]
Please find the requested file attached here - [screen_recording.zip](https://github.com/flutter/flutter/files/5154561/Screen.Recording.2020-09-01.at.12.51.56.PM.zip)
Also, I'd like to mention I'm running **macOS Big Sur 11 beta 5 (20A5354i)** and **Xcode 12.0 Build version 12A8189h**.

**Comment by darshankawar** [comment]
Thanks @kartikeybhardwaj for mentioning the os and xcode version. I am on one version below from this config and you may be seeing the issue due to latest config. Keeping this issue open for further analysis.

**Comment by kartikeybhardwaj** [comment]
I tested more, I reduced Xcode version to **Xcode 11.7 Build version 11E801a**, which is the currently stable. It's not about just `Expanded` or `Flexible` widget, I believe gestures are broken on **Big Sur**. I tried switching to master channel as well, same issue. Root cause is still unknown to me.
Here is output of `flutter run -d macos --verbose` - [log-verbose.txt](https://github.com/flutter/flutter/files/5162366/log-verbose.txt)

**Comment by cbenhagen** [comment]
@kartikeybhardwaj have you found a workaround for this issue?

**Comment by kartikeybhardwaj** [comment]
@cbenhagen Yes, I reverted to macOS *Catalina*. It works here.

**Comment by cbenhagen** [comment]
So far I found out that only `onTapUp` is not firing. When it fails on a certain offset it will also fail on subsequent attempts at that offset. After using the systems zoom gesture once (enable in System Preferences -> Accessibility -> Zoom) it does work for the rest of the users session. You can close an re-open your app and it still works. To reproduce you have to log out and back in again. I have no idea yet why, but @kartikeybhardwaj you could use this as a workaround for now.

Here is the code I tested with:

```dart
import 'dart:ui';

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Color color = Colors.blue;
  Offset lastOffset;
  List<Offset> failed = [];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Center(
          child: GestureDetector(
        onTapDown: (details) {
          if (color == Colors.red && !failed.contains(details.localPosition)) {
            failed.add(lastOffset);
          }
          setState(() {
            color = Colors.red;
          });
          print('tap down '
              'Global: ${details.globalPosition} '
              'Local: ${details.localPosition}');
          lastOffset = details.localPosition;
        },
        onTapUp: (details) {
          setState(() {
            color = Colors.blue;
          });
          print('tap up '
              'Global: ${details.globalPosition} '
              'Local: ${details.localPosition}');
        },
        child: CustomPaint(
          foregroundPainter: FailedPainter(failed),
          child: Container(
            color: color,
            width: 200,
            height: 150,
          ),
        ),
      )),
    );
  }
}

class FailedPainter extends CustomPainter {
  FailedPainter(this.failed);

  final List<Offset> failed;

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawPoints(
        PointMode.points,
        failed,
        Paint()
          ..color = Colors.white
          ..strokeWidth = 1);
  }

  @override
  bool shouldRepaint(FailedPainter oldDelegate) => true;
}
```

**Comment by maxim-saplin** [comment]
Same issue on Big Sur Beta 9, occasional clicks work. Curios fact, same issue seem to be presen in Opera, sometimes I can't close Tabs by clicking on close button

**Comment by thmatuza** [comment]
I have seen the same issue on Flutter for Web when I updated my Mac to Big Sur yesterday.

**Comment by thmatuza** [comment]
Workaround  (enable in System Preferences -> Accessibility -> Zoom) also works for Flutter for Web.

**Comment by ibrahimyu** [comment]
Confirmed, zooming in and out once fixes it for now. Thanks, I thought I had to spend another day downgrading to Catalina...

**Comment by technolion** [comment]
Since Bug Sur is now pushed to the general public, this is a major issue for all Flutter apps on macOS. Clicking on `RaisedButton`, `FlatButton`, `GestureDetector` does not work reliably under Big Sur.
How about raising the priority here?

**Comment by maxim-saplin** [comment]
Having the priority raised and issue fixed before Big Sur becoming GA would be a nice example of project management. The date of Big Sur release was known in summer, as well as this issue.

**Comment by stuartmorgan** [comment]
> Having the priority raised and issue fixed before Big Sur becoming GA would be a nice example of project management. The date of Big Sur release was known in summer, as well as this issue.

Flutter for macOS is currently in alpha, and we explicitly recommend against shipping apps with it at this stage. In addition, there was evidence (from you, in fact) that suggested that this could well be a bug in the Big Sur beta itself—as it apparently affected completely unrelated software—rather than a Flutter issue.

Passive-aggressive comments about project management due to not making a bug in alpha software on a beta OS a top priority is not constructive.

**Comment by stuartmorgan** [comment]
@gw280 When you have your Big Sur environment set up, can you inspect the event stream here? It's a pretty straightforward passthrough, so it may be something subtle in the conversion (e.g., maybe something like details of event timestamps?)

**Comment by AndroidNils** [comment]
Some of our customers confirmed this issue. They press on a button, but nothing happens. After some more tries it worked. Sometime it works with one click (expected behavior), sometimes it works only with multiple clicks (unexpected behavior).

I updated also to Big Sur, but I can't reproduce this issue.

**Comment by gw280** [comment]
I can't reproduce this on my Big Sur machine running 11.0.1 (20B29).

**Comment by stuartmorgan** [comment]
It's odd that only some people see it. For those who are seeing it, is it happening with a mouse, a trackpad, or both (on the off-chance that somehow that is relevant)?

**Comment by ezorrio** [comment]
> It's odd that only some people see it. For those who are seeing it, is it happening with a mouse, a trackpad, or both (on the off-chance that somehow that is relevant)?

Experience same issue since (not quite sure exactly) beta 4 of Big Sur

Occurs both on mouse and touchpad. MacBook Pro 13 (2019), if it makes difference.

**Comment by maxim-saplin** [comment]
MBP 2018, trackpad and mouse, Flutter dev channel, Big Sur stable. Created a blank project with single button and the issue is not reproducible there. In a project with lots of dependencies it is still in place: https://github.com/maxim-saplin/dikt

**Comment by fanchou** [comment]
I reproduce this on my Big Sur machine running 11.0.1 everyday.

**Comment by technolion** [comment]
> It's odd that only some people see it. For those who are seeing it, is it happening with a mouse, a trackpad, or both (on the off-chance that somehow that is relevant)?

I can add something here: The problem occurs after booting macOS Big Sur. Then a flutter app does not respond correctly on mouse clicks. After zooming in/out (CTRL mouse wheel) the problem is gone and stays gone until you reboot macOS again.

**Comment by fanchou** [comment]
This problem does not exist when the window is maximized.

Maybe it's related to Visual Density?

> I reproduce this on my Big Sur machine running 11.0.1 everyday.

**Comment by maxim-saplin** [comment]
Channel dev, 1.24.0-10.2.pre, on macOS 11.1 20C5048k darwin-x64. Below is a video with the issue demonstrated in a small project. At first clicks work, few seconds latter button shows click animation though onclick event is not fired.

![clicks](https://user-images.githubusercontent.com/7947027/99965110-70304f00-2da5-11eb-980e-ed149d57f72d.gif)

**Comment by avdept** [comment]
Same issue with checkboxes. When moving few pixels aside - starts to work fine. Using trackpad
![Screen Recording 2020-11-24 at 20 36 04](https://user-images.githubusercontent.com/1757017/100138560-e19bfa80-2e96-11eb-97ea-08fba197f051.gif)

**Comment by stuartmorgan** [comment]
https://github.com/flutter/flutter/issues/70529#issuecomment-733330469 has details about what's happening here.

We'll need to look at the native side of the event stream, since it sounds like we're sending bad events at the embedding layer.

**Comment by gw280** [comment]
Seeing as we have higher priority items to deal with right now, I've filed a follow up issue to investigate the embedding event stream, and I am going to advocate for the framework workaround in https://github.com/flutter/flutter/issues/70529#issuecomment-733330469 to resolve this particular issue for now.

**Comment by csells** [comment]
In my opinion, this all but makes macOS desktop development on Big Sur impossible. Here's another repro case if needed:

```dart
class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: Text(App.title)),
        body: Row(
          children: [
            Expanded(
              child: ListView(
                children: [
                  for (final s in ['one', 'two', 'three'])
                    ListTile(
                      onTap: () => print('ListTile.onTap'),
                      title: Text(s),
                    )
                ],
              ),
            ),
            Expanded(
              child: Container(
                child: Center(
                  child: OutlineButton(
                    onPressed: () => print('TextButton.onPressed'),
                    child: Text('click'),
                  ),
                ),
              ),
            ),
          ],
        ),
      );
}
```

You can see when I run this app on my Big Sur 11.0.1 machine that keypress events are always registered but mouse clicks are intermittent:

![Dec-06-2020 10-32-03](https://user-images.githubusercontent.com/2568253/101289189-c9d64780-37af-11eb-9b50-9b21b99980a3.gif)

Here's my flutter doctor output:

```sh
❯ flutter doctor -v                        
[✓] Flutter (Channel master, 1.25.0-5.0.pre.91, on macOS 11.0.1 20B29 darwin-x64, locale en-US)
    • Flutter version 1.25.0-5.0.pre.91 at /Users/csells/flutter
    • Framework revision 7dade22f25 (3 hours ago), 2020-12-06 21:34:04 +0530
    • Engine revision 6491c7518f
    • Dart version 2.12.0 (build 2.12.0-120.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/csells/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.0.1 20B29 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 86.0.4240.198

• No issues found!
```

**Comment by csells** [comment]
@cbracken fyi

**Comment by eikob** [comment]
It seems the same issue that has been solved for the web in #70529.

Do we really need to reinvent the wheel here? Seems to be related to HighDPI displays. Clicks wok always on some pixels / mouse positions, and on others don't.

Makes development (and potentially deploying) impossible.

**Comment by stuartmorgan** [comment]
> Do we really need to reinvent the wheel here?

The fixes will be similar, but not identical. The macOS embedding and the web embedding are not the same code.

**Comment by cbracken** [comment]
Patch is out to fix this on desktop. Adding tests.

**Comment by cbracken** [comment]
This has landed in the engine, it'll be fixed in the framework once it rolls the the framework -- this typically happens within a day.

**Comment by AndroidNils** [comment]
Yeah! Thank you! 🚀

**Comment by cbracken** [comment]
This has now rolled to the framework on master branch.

**Comment by mkulesh** [comment]
Thank you! I have now tested it with master channel, 1.26.0-2.0.pre.172, on macOS 11.0.1 20B29 darwin-x64, locale en-GB. It works now as desired, all clicks now work.

**Comment by spkersten** [comment]
Will this also be fixed on 1.22? We use macOS builds of our apps very often during development to be able to check layouts at different sizes and for general speed (Simulator tend to get sluggish).
